internal/metrics/binpack.go

```diff
@@ -0,0 +1,117 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"math"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/binpack"
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+	"github.com/palantir/pkg/metrics"
+)
+
+type reportableEfficiency struct {
+	CPU    float64
+	Memory float64
+	GPU    float64
+}
+
+func reportableEfficiencyFromPackingEfficiency(efficiency *binpack.PackingEfficiency) reportableEfficiency {
+	return reportableEfficiency{
+		CPU:    efficiency.CPU,
+		Memory: efficiency.Memory,
+		GPU:    efficiency.GPU,
+	}
+}
+
+func reportableEfficiencyFromAvgPackingEfficiency(efficiency *binpack.AvgPackingEfficiency) reportableEfficiency {
+	return reportableEfficiency{
+		CPU:    efficiency.CPU,
+		Memory: efficiency.Memory,
+		GPU:    efficiency.GPU,
+	}
+}
+
+const (
+	packingEfficiencyMetricName = "foundry.spark.scheduler.packing_efficiency"
+
+	packingResourceTagKey = "foundry.spark.scheduler.packing_resource"
+	cpuTagValue           = "CPU"
+	memoryTagValue        = "Memory"
+	gpuTagValue           = "GPU"
+	maxTagValue           = "Max"
+
+	packingEfficiencyFunctionNameTagKey = "foundry.spark.scheduler.packing_function"
+
+	// non-existing node to explicitly report average packing efficiency
+	avgEfficiencyNodeName = "average"
+)
+
+var (
+	cpuTag    = metrics.MustNewTag(packingResourceTagKey, cpuTagValue)
+	memoryTag = metrics.MustNewTag(packingResourceTagKey, memoryTagValue)
+	gpuTag    = metrics.MustNewTag(packingResourceTagKey, gpuTagValue)
+	// represents higher of CPU and Memory packing efficiencies. GPU is explicitly excluded for now
+	maxTag = metrics.MustNewTag(packingResourceTagKey, maxTagValue)
+)
+
+// ReportPackingEfficiency report packing efficiency metrics for a single packing result.
+func ReportPackingEfficiency(
+	ctx context.Context,
+	packingFunctionName string,
+	nodesSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	packingResult *binpack.PackingResult) {
+
+	contextTags := metrics.TagsFromContext(ctx)
+	registry := metrics.NewRootMetricsRegistry()
+
+	packingFunctionTag := metrics.MustNewTag(packingEfficiencyFunctionNameTagKey, packingFunctionName)
+
+	// report packing efficiencies per node
+	for _, efficiency := range packingResult.PackingEfficiencies {
+		hostTag := HostTag(ctx, efficiency.NodeName)
+		emitMetrics(registry, contextTags, hostTag, packingFunctionTag, reportableEfficiencyFromPackingEfficiency(efficiency))
+	}
+
+	// report avg packing efficiency for all nodes at once
+	efficiency := computeAvgPackingEfficiencyForResult(nodesSchedulingMetadata, packingResult)
+	avgHostTag := HostTag(ctx, avgEfficiencyNodeName)
+	emitMetrics(registry, contextTags, avgHostTag, packingFunctionTag, reportableEfficiencyFromAvgPackingEfficiency(&efficiency))
```

Comment:
Im a bit confused as to why we're reporting twice here?

Reply:
This is the pattern we've used in scaler telemetry: report efficiency per node and then separately for all nodes at once. The upside of reporting by node is that we can then see what the efficiency in every node. Obviously, comes with an increase in amount of data. We could start with just the avg efficiency, disregarding node stats.